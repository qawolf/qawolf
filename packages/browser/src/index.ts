// public API
export { BrowserContext, register } from "./ContextManager";
export { Page } from "./page/PageManager";
// TODO move to jest-playwright
export { launch, LaunchResult } from "./launch";

// internal API
export * from "./actions";
export * from "./keyboard";
export { ContextManager } from "./ContextManager";
export { PageManager } from "./page/PageManager";
