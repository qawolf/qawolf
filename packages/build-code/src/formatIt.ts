import { Step } from "@qawolf/types";
import { decodeHtml } from "@qawolf/web";

export const formatDescription = (step: Step): string => {
  const target = step.html.node;

  const attrs = target.attrs || {};

  // ex. "departure date"
  let description =
    attrs.labels ||
    attrs.name ||
    attrs.placeholder ||
    decodeHtml(attrs.innertext) ||
    attrs.id ||
    attrs.alt ||
    "";

  description = description.trim();

  if (description.length > 40) {
    description = `${description.substring(0, 40)}...`;
  }

  if (description.length) {
    // strip single quotes so they don't mess up formatting
    return ` "${description}"`.replace("'", "");
  }

  return "";
};

export const formatIt = (step: Step): string => {
  if (step.action === "scroll") {
    return `can scroll`;
  }

  const description = formatDescription(step);

  const target = step.html.node;

  // link/input
  const tagName = `${target.name === "a" ? " link" : ` ${target.name}` || ""}`;

  if (step.action === "click") {
    return `can click${description}${tagName}`;
  }

  if (step.action === "select") {
    return `can select${description}`;
  }

  if (step.action === "type") {
    if (!step.value) {
      return `can clear${description}${tagName}`;
    }

    const value = step.value as string;
    if (value.indexOf("↓Enter") === 0) {
      return `can Enter`;
    }

    if (value.indexOf("↓Tab") === 0) {
      return `can Tab`;
    }

    return `can type into${description}${tagName}`;
  }

  throw new Error(`Invalid step action ${step.action}`);
};
