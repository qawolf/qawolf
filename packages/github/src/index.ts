import Octokit from "@octokit/rest";

const octokit = new Octokit({
  auth: "d9e1b19607b2b07ed5c6f2e78fa5fb1bd3cabf26"
});

const main = async () => {
  const res = await octokit.checks.create({
    owner: "qawolf",
    repo: "internal",
    name: "QA Wolf",
    head_sha: "9c52a7a6f2090c58bcb23ae066ce31920789f71c"
  });

  console.log("RESPONSE", res);
};

main();
