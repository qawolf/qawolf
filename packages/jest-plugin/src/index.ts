// setup fail-fast since we do not want to continue when a step fails
// hopefully jest-circus implements this behavior, but was not the case when I tested jest 24.9.0
// https://github.com/facebook/jest/issues/2867#issuecomment-476647864
// https://github.com/facebook/jest/issues/6527#issuecomment-506278251

type Callback = (name: string) => void;

const j = jasmine as any;

// require since there are no types
const failFast = require("@qawolf/jasmine-fail-fast");
const jasmineEnv = j.getEnv();
jasmineEnv.addReporter(failFast.init());
