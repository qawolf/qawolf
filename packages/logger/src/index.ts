import { Logger } from "./Logger";

export const browserLogger = new Logger("browser");

export const logger = new Logger("server");
