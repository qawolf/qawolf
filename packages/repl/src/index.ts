import { Reporter } from "./Reporter";

// export for types to be happy
export { repl, ReplWithContext } from "./ReplWithContext";
export { CONTEXT } from "./Context";

import { repl } from "./ReplWithContext";
import { CONTEXT } from "./Context";

// export with module.exports for Jest to be happy
const mainExport = Reporter as any;
mainExport.CONTEXT = CONTEXT;
mainExport.repl = repl;
module.exports = mainExport;
