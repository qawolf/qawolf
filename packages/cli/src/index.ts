#!/usr/bin/env node

import { CONFIG } from "@qawolf/config";
import { logger } from "@qawolf/logger";
import { yellow } from "kleur";
import program from "commander";
import { snakeCase } from "lodash";
import { saveCiTemplate } from "./ci";
import { record } from "./record";
import { test } from "./test";
import { parseUrl } from "./utils";
const { version } = require("../package");

program.version(version);

program.usage("<command> [options]");

let recordCommand = program
  .command("record <url> [name]")
  .description("record a workflow and create a test");

if (CONFIG.development) {
  recordCommand = recordCommand.option(
    "-e, --events",
    "save events (for debugging)"
  );
}

recordCommand.action(async (urlArgument, optionalName, cmd) => {
  const url = parseUrl(urlArgument);
  logger.verbose(`record url "${url.href}"`);

  const name = snakeCase(optionalName || url.hostname!);
  await record(url, name, cmd.events);
});

program
  .command("test [name]")
  .description("run a test")
  .action(async optionalName => {
    await test(optionalName);
  });

program
  .command("azure")
  .description("set up an Azure Pipeline")
  .action(async () => {
    await saveCiTemplate("azure");
  });

program
  .command("circleci")
  .description("set up CircleCI")
  .action(async () => {
    await saveCiTemplate("circleci");
  });

program
  .command("github")
  .description("set up a GitHub Action")
  .action(async () => {
    await saveCiTemplate("github");
  });

program
  .command("gitlab")
  .description("set up GitLab CI/CD")
  .action(async () => {
    await saveCiTemplate("gitlab");
  });

program.arguments("<command>").action(cmd => {
  console.log(yellow(`Invalid command "${cmd}"\n`));
  program.help();
});

program.allowUnknownOption(false);

program.parse(process.argv);

if (!process.argv.slice(2).length) {
  console.log("\n");
  program.outputHelp();
}
