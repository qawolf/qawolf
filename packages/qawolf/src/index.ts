#!/usr/bin/env node
import updateNotifier from "update-notifier";

const pkg = require("../package.json");

const isCLI = !module.parent;
if (isCLI) {
  updateNotifier({ pkg }).notify();

  require("@qawolf/cli");
}

// export public API
export { BrowserContext, launch, Page } from "@qawolf/browser";

export { repl } from "@qawolf/repl";

export { sleep, waitFor, waitUntil } from "@qawolf/web";

// make repl a global
import { repl } from "@qawolf/repl";
(global as any).repl = repl;
