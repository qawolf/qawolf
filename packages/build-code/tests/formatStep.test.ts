import { Step } from "@qawolf/types";
import { formatStep } from "../src/formatStep";

describe("formatStep", () => {
  describe("click", () => {
    it("formats click on link step", () => {
      const step: Step = {
        action: "click",
        html: {
          ancestors: [],
          node: {
            attrs: { innertext: "contact" },
            name: "a",
            type: "tag"
          }
        },
        index: 0,
        page: 0
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });

    it("formats click on submit input step", () => {
      const step: Step = {
        action: "click",
        html: {
          ancestors: [],
          node: {
            attrs: { innertext: "someone's" },
            name: "input",
            type: "tag"
          }
        },
        index: 1,
        page: 0
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });
  });

  describe("scroll", () => {
    it("formats scroll action", () => {
      const step: Step = {
        action: "scroll",
        html: {
          ancestors: [],
          node: {
            name: "html",
            type: "tag"
          }
        },
        index: 0,
        page: 0,
        value: { x: 0, y: 10 }
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });
  });

  describe("select", () => {
    it("formats select action", () => {
      const step: Step = {
        action: "select",
        html: {
          ancestors: [],
          node: {
            attrs: {
              name: "select1"
            },
            name: "select",
            type: "tag"
          }
        },
        index: 0,
        page: 0,
        value: "spirit"
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });
  });

  describe("type", () => {
    it("formats clear an input", () => {
      const step: Step = {
        action: "type",
        html: {
          ancestors: [],
          node: {
            name: "input",
            type: "tag"
          }
        },
        index: 0,
        page: 0,
        value: null
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });

    it("formats type into text input", () => {
      const step: Step = {
        action: "type",
        html: {
          ancestors: [],
          node: {
            attrs: {
              id: "input1",
              name: "username",
              placeholder: "Jane Doe"
            },
            name: "input",
            type: "tag"
          }
        },
        index: 0,
        page: 0,
        value: "spirit"
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });

    it("formats type into password input", () => {
      const step: Step = {
        action: "type",
        html: {
          ancestors: [],
          node: {
            attrs: {
              id: "input2",
              placeholder: "secret"
            },
            name: "input",
            type: "tag"
          }
        },
        index: 10,
        page: 0,
        value: "supersecret"
      };

      const formattedStep = formatStep(step);
      expect(formattedStep).toMatchSnapshot();
    });
  });
});
