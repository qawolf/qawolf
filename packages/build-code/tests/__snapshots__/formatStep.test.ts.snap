// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatStep click formats click on link step 1`] = `
Object {
  "it": "can click \\"contact\\" link",
  "method": "browser.click(selectors[0])",
}
`;

exports[`formatStep click formats click on submit input step 1`] = `
Object {
  "it": "can click \\"someones\\" input",
  "method": "browser.click(selectors[1])",
}
`;

exports[`formatStep scroll formats scroll action 1`] = `
Object {
  "it": "can scroll",
  "method": "browser.scroll(selectors[0], { x: 0, y: 10 })",
}
`;

exports[`formatStep select formats select action 1`] = `
Object {
  "it": "can select \\"select1\\"",
  "method": "browser.select(selectors[0], \\"spirit\\")",
}
`;

exports[`formatStep type formats clear an input 1`] = `
Object {
  "it": "can clear input",
  "method": "browser.type(selectors[0], null)",
}
`;

exports[`formatStep type formats type into password input 1`] = `
Object {
  "it": "can type into \\"secret\\" input",
  "method": "browser.type(selectors[10], \\"supersecret\\")",
}
`;

exports[`formatStep type formats type into text input 1`] = `
Object {
  "it": "can type into \\"username\\" input",
  "method": "browser.type(selectors[0], \\"spirit\\")",
}
`;
